### PR TITLE
EN-65714: move release tagging

### PR DESCRIPTION
**Pull request by sarah-salinas**
This commit moves the release tagging after a successful docker build. In order to do this, it separates the docker build from the publishing of the images into separate Jenkinsfile steps.

### CODE REVIEW

**Comment by sarah-salinas** [comment]
## Testing

### Simulated Main Build

https://jenkins-build.socrata.com/job/data-coordinator-branch/31/console

### Release Dry Run

TODO

### Simulated Release Build

TODO

**Review by rjmac — APPROVED** [review]

File: Jenkinsfile [hunk 1]
```diff
@@ -72,56 +83,63 @@ pipeline {
...
+        success {
+          script {
+            if (params.RELEASE_BUILD){
+              env.GIT_TAG = releaseTag.getFormattedTag(params.RELEASE_NAME)
+              if (releaseTag.doesReleaseTagExist(params.RELEASE_NAME)) {
+                echo "REBUILD: Tag ${env.GIT_TAG} already exists"
+                return
+              }
+              if (params.RELEASE_DRY_RUN) {
+                echo "DRY RUN: Would have created ${env.GIT_TAG} and pushed it to the repo"
```
Comment (ds2643) [inline Jenkinsfile:114]: Again, I don't think it's possible to reach this statement.

File: Jenkinsfile [hunk 2]
```diff
@@ -146,16 +164,15 @@ pipeline {
...
+          marathonDeploy(serviceName: env.DEPLOY_PATTERN, tag: env.DOCKER_TAG, environment: 'staging')
         }
       }
     }
   }
   post {
     failure {
       script {
-        if (!isPr) {
+        if (env.JOB_NAME == "${service}/main") {
```
Comment (ds2643) [inline Jenkinsfile:175]: Did isPR stop working?

**Review by ds2643 — COMMENTED** [review]
Minor comments left. I want to spend time understanding the bigger picture.

File: Jenkinsfile (hunk 2 above)
Comment (sarah-salinas) [reply Jenkinsfile:175]: No, but it's not specific enough.  When I am testing on a branch, I don't want the build failures to post to teams.  We only want the main build failures to post to teams.  Any release build failures will show up in the "Release-Build" job, so we'll see them there.

File: Jenkinsfile [hunk 3]
```diff
@@ -37,19 +38,29 @@ pipeline {
     WEBHOOK_ID = 'WEBHOOK_IQ'
   }
   stages {
-    stage('Release Tag') {
+    stage('Publish Library') {
```
Comment (sarah-salinas) [inline Jenkinsfile:41]: I made a ticket to automate the publishing of the library when there's a change to the verison.sbt file
https://socrata.atlassian.net/browse/EN-65991

**Comment by sarah-salinas** [comment]
## Testing

### Simulated Main Build

https://jenkins-build.socrata.com/job/data-coordinator-branch/34/console - 🟢 

Stages
- Build
- Docker Build
- Publish Image: internal only
- Deploy: staging

Description: `data-coordinator:STAGING_34_f52ef745`

### Release Dry Run

https://jenkins-build.socrata.com/job/data-coordinator-branch/35/console - 🟢 

Stages
- Build
- Docker Build

Description: `data-coordinator:H23 - DRY RUN`

### Simulated Release Build

https://jenkins-build.socrata.com/job/data-coordinator-branch/36/console - 🟢 

Stages
- Build
- Docker Build , Post: release tag
- Publish Image: all environments, Post: rms create build

Description: `data-coordinator:H23`

**Review by rjmac — APPROVED** [review]

File: Jenkinsfile [hunk 4]
```diff
@@ -60,71 +73,91 @@ pipeline {
...
+            )
+          }
+        }
+      }
+      post {
+        success {
+          script {
+            if (params.RELEASE_BUILD){
+              env.GIT_TAG = releaseTag.getFormattedTag(params.RELEASE_NAME)
+              if (releaseTag.doesReleaseTagExist(params.RELEASE_NAME)) {
```
Comment (ds2643) [inline Jenkinsfile:117]: I would lift these statements into mutually exclusive conditions.

Comment (ds2643) [reply Jenkinsfile:117]:
```groovy
if (params.RELEASE_BUILD){
	env.GIT_TAG = releaseTag.getFormattedTag(params.RELEASE_NAME)
        if (releaseTag.doesReleaseTagExist(params.RELEASE_NAME)) {
        echo "REBUILD: Tag ${env.GIT_TAG} already exists"
	return
      }
      else if (params.RELEASE_DRY_RUN) {
	echo "DRY RUN: Would have created ${env.GIT_TAG} and pushed it to the repo"
	currentBuild.description = "${service}:${params.RELEASE_NAME} - DRY RUN"
      } else {
      	releaseTag.create(params.RELEASE_NAME)
	}
}

```

File: Jenkinsfile (hunk 2 above)
Comment (ds2643) [reply Jenkinsfile:194]: For clarity, I'd express as such:
```groovy
boolean buildingMain = env.JOB_NAME = "${service}/main"
if (buildingMain) {...}
```

**Review by ds2643 — APPROVED** [review]
I left some minor comments about style, but this pull-request otherwise closely resembles previously approved work on other repositories.

File: Jenkinsfile (hunk 2 above)
Comment (sarah-salinas) [reply Jenkinsfile:194]: Hmm, I'll think about this suggestion for future PRs.

File: Jenkinsfile (hunk 4 above)
Comment (sarah-salinas) [reply Jenkinsfile:117]: I'll think about this for future PRs.  
I think the `if return` pattern can be more extensible that `if else if else`.

**Pull request merged**